### PR TITLE
[Docs] Explain metric internals

### PR DESCRIPTION
**Pull request by SkafteNicki**
## What does this PR do?

Fixes https://github.com/PyTorchLightning/pytorch-lightning/issues/4921
Fixes https://github.com/PyTorchLightning/pytorch-lightning/issues/5737

Some users have complained that some side-effects of metrics is not well documented. First issue relates to ``update`` being called twice during ``forward`` (nessesary as we are both computing local statistics and updating global statistics). Second issue relates to the result of ``compute`` being internally cached (needed for logging the metric). This PR tries to document this behaviour.

cc: @edenlightning 

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified
 - [x] **Check that target branch and milestone match!**

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Review by justusschock — APPROVED** [review]

**Review by Borda — APPROVED** [review]

File: docs/source/extensions/metrics.rst [hunk 1]
```diff
@@ -258,6 +258,50 @@ In practise this means that:
...
+----------
+
+.. autoclass:: pytorch_lightning.metrics.Metric
+    :noindex:
+
+Internal implementation details
+-------------------------------
+
+This section briefly describe how metrics work internally. We encourage looking at the source code for more info.
+Internally, Lightning wraps the user defined ``update()`` and ``compute()`` method. We do this to among other to
```
Comment (carmocca) [inline docs/source/extensions/metrics.rst:272]: "We do this to among other to"

Mind rephrasing?

File: docs/source/extensions/metrics.rst [hunk 2]
```diff
@@ -258,6 +258,50 @@ In practise this means that:
...
+Simiarly, calling ``compute()`` does the following internally
+
+1. Syncs metric states between processes
+2. Reduce gathered metric states
+3. Calls the user defined ``compute()`` method on the gathered metric states
+4. Cache computed result
+
+From a user's standpoint this has one important side-effect: computed results are cached. This means that no
+matter how many times ``compute`` is called after one and another, it will continue to return the same result.
+The cashe is first emptied on the next call to ``update``.
```
Comment (carmocca) [inline docs/source/extensions/metrics.rst:288]:
```suggestion
The cache is first emptied on the next call to ``update``.
```

File: docs/source/extensions/metrics.rst [hunk 3]
```diff
@@ -258,6 +258,50 @@ In practise this means that:
...
+
+``forward`` serves the dual purpose of both returning the metric on the current data and updating the internal
+metric state for accumulating over multiple batches. The ``forward()`` method achives this by combining calls
+to ``update`` and ``compute`` in the following way (assuming metric is initialized with ``compute_on_step=True``):
+
+1. Calls ``update()`` to update the global metric states (for accumulation over multiple batches)
+2. Caches the globel state
```
Comment (carmocca) [inline docs/source/extensions/metrics.rst:295]:
```suggestion
2. Caches the global state
```

File: docs/source/extensions/metrics.rst [hunk 4]
```diff
@@ -258,6 +258,50 @@ In practise this means that:
...
+3. Calls ``reset()`` to clear global metric state
+4. Calls ``update()`` to update local metric state
+5. Calls ``compute()`` to calculate metric for current batch
+6. Restores the global state
+
+This structure has the important consequence the user defined ``update`` is called **twice** during a single
```
Comment (carmocca) [inline docs/source/extensions/metrics.rst:301]:
```suggestion
This procedure has the consequence of calling the user defined ``update`` **twice** during a single
```

File: docs/source/extensions/metrics.rst (hunk 1 above)
Comment (SkafteNicki) [reply docs/source/extensions/metrics.rst]: Done :]

**Review by carmocca — APPROVED** [review]

**Review by awaelchli — APPROVED** [review]

**Review by teddykoker — APPROVED** [review]

**Pull request merged**